templates/validator/charts.html

```diff
@@ -154,111 +154,149 @@
+                name: "Execution Income",
+                data: executionIncomeHistory,
+                showInNavigator: false
+              },
+              {
+                name: "Consensus Income",
+                data: incomeHistory,
+                showInNavigator: true
+              }
+            ],
+            tooltip: {
+              split: false,
+              shared: true,
+              formatter: (tooltip) => {
+                var text = ``
+                var total = 0
 
-                                // add total if hovered point contains rewards for both EL and CL
-                                if (tooltip.chart.hoverPoints.length > 1) {
-                                    text += `<b>Total:</b> ${getIncomeChartValueString(total, currency, ethPrice)}`
-                                }
+                // date and epochs
+                const startEpoch = timeToEpoch(tooltip.chart.hoverPoint.x)
+                const timeForOneDay = 24 * 60* 60 * 1000
+                const endEpoch = timeToEpoch(tooltip.chart.hoverPoint.x + timeForOneDay) - 1
+                const startDate = luxon.DateTime.fromMillis(tooltip.chart.hoverPoints[0].x)
+                const endDate = luxon.DateTime.fromMillis(epochToTime(endEpoch+1))
+                text += `${startDate.toFormat("MMM-dd-yyyy HH:mm:ss")} - ${endDate.toFormat("MMM-dd-yyyy HH:mm:ss")}<br> Epochs ${startEpoch} - ${endEpoch}<br/>`;
 
-                                return text
-                            }
-                        }
-                    });
-                })
+                // income
+                for (var i = 0; i < tooltip.chart.hoverPoints.length; i++) {
+                  const value = tooltip.chart.hoverPoints[i].y;
+                  text += `<span style="color:${tooltip.chart.hoverPoints[i].series.color}">\u25CF</span>  <b>${tooltip.chart.hoverPoints[i].series.name}:</b> ${getIncomeChartValueString(value, currency, ethPrice)}<br/>`
+                  total += value
+                }
 
-        </script>
+                // add total if hovered point contains rewards for both EL and CL
+                if (tooltip.chart.hoverPoints.length > 1) {
+                  text += `<b>Total:</b> ${getIncomeChartValueString(total, currency, ethPrice)}`
+                }
+
+                return text
+              }
+            }
+          });
+
+          var incomeChartLoaded = false;
+          $("#load-income-btn").on("click", () => {
+            if (incomeChartLoaded) {
+              return
+            }
+            incomeChartLoaded = true
+            const minTime = incomeHistory[0].x / 1000
+            const genesisTimestamp = {{$.ChainGenesisTimestamp}}
+            const minDay = Math.floor((minTime - genesisTimestamp) / (24 * 60 * 60)) // floor shouldn't be necessary but just in case
+
+            const url = '/dashboard/data/allbalances?validators=' + {{.ValidatorIndex}} + '&end_day=' + Math.max(0, minDay - 1)
```

Comment:
It is kind of odd to use a dashboard function on the validator page.

Have you considered renaming `func DashboardDataBalanceCombined` to something more generic (i.e. not being labeled as something for the Dashboard only) and adding another route that calls the exact same function?

Right now, there is some inconsistency: When the chart data is loaded initially (with the limited timeframe), it is done WITHOUT `func DashboardDataBalanceCombined` by using `GetValidatorIncomeHistoryChart` and `getExecutionChartData`. Both are used in `DashboardDataBalanceCombined` to but it feels error prone (at least a bit).

Please share your thoughts on this topic.

Just a reminder, changing the endpoint itself is complicated as the app relies on it too.

Reply:
**Regarding the endpoint:**
Yes I have considered it. IMO there shouldn't be a "dashboard endpoint" in the first place, as the dashboard does not identify the resources fetched by the endpoint (and having non-identifying parts in the URI leads to issues like this). Considering that the app also uses this endpoint and that another endpoint would just do the exact same thing, I have opted to use the "dashboard enpoint" as it is.

**Regarding DashboardDataBalanceCombined:**
I don't see the inconsistency. `GetValidatorIncomeHistoryChart` and `getExecutionChartData` are functions that fetch data. `DashboardDataBalanceCombined` is a handler function. `validator.go` has a handler function. Both handler functions use `GetValidatorIncomeHistoryChart` and `getExecutionChartData` to fetch data. I can't use `DashboardDataBalanceCombined` inside the validator handler because `DashboardDataBalanceCombined` responds with a JSON payload for each HTTP request.

**Verdict:**
Won't do.